File: examples/canonical-analytics/python/src/analytics_workbench/charts/heatmap_chart.py
import plotly.graph_objects as go
from ..analytics.heatmap import EventHeatmap
from datetime import datetime

def build_heatmap_figure(agg: EventHeatmap) -> dict:
    fig = go.Figure()
    
    total_count = sum(sum(row) for row in agg.counts) if agg.counts else 0
    
    if not agg.days:
        fig.add_annotation(text="No data available", showarrow=False, font=dict(size=20))
        fig.update_layout(title_text="Event Heatmap", xaxis_title_text="Hour of Day", yaxis_title_text="Day of Week")
    else:
        fig.add_trace(go.Heatmap(
            z=agg.counts,
            x=agg.hours,
            y=agg.days,
            colorscale="Blues",
            name="Heatmap",
            hovertemplate="Day: %{y}<br>Hour: %{x}<br>Count: %{z}<extra></extra>"
        ))
        
        fig.update_layout(
            title_text="Event Heatmap",
            xaxis_title_text="Hour of Day",
            yaxis_title_text="Day of Week"
        )
        
    fig_dict = fig.to_dict()
    fig_dict["meta"] = {
        "total_count": total_count,
        "filters_applied": [],
        "generated_at": datetime.utcnow().isoformat()
    }
    return fig_dict

File: examples/canonical-analytics/rust/src/charts/heatmap_chart.rs
use super::to_figure_json;
use crate::analytics::EventHeatmap;
use plotly::common::{Title, ColorScalePalette};
use plotly::{HeatMap, Plot};
use serde_json::Value;

pub fn build_heatmap_figure(agg: &EventHeatmap) -> Value {
    let mut plot = Plot::new();

    if agg.counts.is_empty() || agg.counts[0].is_empty() {
        let trace = HeatMap::new(vec![0.0], vec![0.0], vec![vec![0.0]]);
        plot.add_trace(trace);
        let layout = plotly::Layout::new().title(Title::with_text("No data for selected filters"));
        plot.set_layout(layout);
        return to_figure_json(&plot, 0, true);
    }

    // counts is 7 days x 24 hours. HeatMap takes Z as Vec<Vec<N>>
    let mut z_data = Vec::new();
    for day_row in &agg.counts {
        let mut row_f64 = Vec::new();
        for &val in day_row {
            row_f64.push(val as f64);
        }
        z_data.push(row_f64);
    }

    let total_count: i64 = agg.counts.iter().map(|row| row.iter().sum::<i64>()).sum();

    // HeatMap X is hours, Y is days
    let trace = HeatMap::new(agg.hours.clone(), agg.days.clone(), z_data)
        .color_scale(ColorScalePalette::Blues.into());
    
    plot.add_trace(trace);

    let layout = plotly::Layout::new()
        .title(Title::with_text("Activity Heatmap"))
        .x_axis(plotly::layout::Axis::new().title(Title::with_text("Hour of Day")))
        .y_axis(plotly::layout::Axis::new().title(Title::with_text("Day of Week")));
    plot.set_layout(layout);

    to_figure_json(&plot, total_count, true)
}